### DISCUSSION TITLE
vue 自定义组件 使用RecursionField组件 报错

### DISCUSSION THREAD

**Discussion by marspal** [alibaba/formily#2981] [Q&A 2.x]
![image](https://user-images.githubusercontent.com/41223123/160232697-a6750fc8-6c6d-46d6-b98e-1497c41a05e4.png)

**Comment by Gyanxie** [comment]
最近也有遇到类似问题。 即依赖的某个npm包，有多个版本的情况下， webpack打包打包的版本和import的版本不一致， 可能会出现import 的方法会undifined 的问题。

### 例子
-  @formily/shared和html-minifier同时依赖了param-case包
- package-lock.json
```

     "@formily/shared": {
          "version": "2.1.11",
          "resolved": "http://registry.npm.oa.com/@formily/shared/download/@formily/shared-2.1.11.tgz",
          "integrity": "sha1-hfWBIiAicD5/Uq2t7FHN4MD/+MQ=",
          "requires": {
            "@formily/path": "2.1.11",
            "camel-case": "^4.1.1",
            "lower-case": "^2.0.1",
            "no-case": "^3.0.4",
            "param-case": "^3.0.4",
            "pascal-case": "^3.1.1",
            "upper-case": "^2.0.1"
          }
        },
   "html-minifier": {
      "version": "3.5.21",
      "resolved": "http://registry.npm.oa.com/html-minifier/download/html-minifier-3.5.21.tgz",
      "integrity": "sha1-0AQOBUcw41TbAIRjWTGUAVIS0gw=",
      "dev": true,
      "requires": {
        "camel-case": "3.0.x",
        "clean-css": "4.2.x",
        "commander": "2.17.x",
        "he": "1.2.x",
        "param-case": "2.1.x",
        "relateurl": "0.2.x",
        "uglify-js": "3.4.x"
      },
```

- formily中使用param-case 3.0.4的版本， 引入的代码
```
import { paramCase } from 'param-case';
```

- 项目中的html-minifier使用了param-case的2.1.x 的版本
- 这两个版本的区别： param-case2是export default function paramCase, 而 param-case3是export.paramCase

- npm i 安装时候，由于顺序问题把param-case安装到了node_modules的根目录， webpack打包时从node_modules目录查找到的是2.x版本的param-case, 于是最后打包的是2.x的版本，导致import { paramCase } from 'param-case'得到的是undefined， 所以导致后续的代码异常。

### 解决办法
- 1. 手动安装param-case3到对应的版本, 让该版本放在node_modeuls根目录，按照项目中的 webpack npm引入规则可以拿到想要的包。
- 2. 修改webpack获取node_modules的顺序， 有个所谓的 resolve属性, 不过怎么改由于我比较菜可以自己了解相关信息。https://zhuanlan.zhihu.com/p/138288718
